aspnet/mvc/controllers/testing.rst

```diff
@@ -1,8 +1,173 @@
-.. include:: /../common/stub-topic.txt
+Testing Controller Logic
+========================
+By `Steve Smith`_
 
-|stub-icon| Testing Controller Logic
-====================================
+Controllers in ASP.NET MVC apps should be small and focused on user-interface concerns. Large controllers that deal with non-UI concerns are more difficult to test and maintain.
 
-.. include:: /../common/stub-notice.txt
+.. contents:: Sections
+	:local:
+	:depth: 1
+	
+`View sample files <https://github.com/aspnet/Docs/tree/master/aspnet/mvc/controllers/testing/sample>`_
 
-.. _issue: https://github.com/aspnet/Docs/issues/123
+What is Controller Logic
+------------------------
+*Controllers* define groups of related *actions*. The grouping of related actions into controllers is useful for :doc:`routing </fundamentals/routing>`, applying :doc:`filters <filters>`, and :doc:`injecting common dependencies <dependency-injection>`. 
+
+:doc:`Learn more about controllers and actions <actions>`.
+
+Controller logic should be minimal and not be focused on business logic or infrastructure concerns (for example, data access). Test controller logic, not the framework. Test how the controller *behaves* based on valid or invalid inputs. Test controller responses based on the result of the business operation it performs.
+
+Typical controller responsibilities:
+	- Verify ``ModelState.IsValid``
+	- Return an error response if ``ModelState`` is invalid
+	- Retrieve a business entity from persistence
+	- Perform an action on the business entity
+	- Save the business entity to persistence
+	- Return an appropriate ``IActionResult``
+
+Unit Testing
+------------
+:doc:`Unit testing </testing/unit-testing>` involves testing a part of an app in isolation from its infrastructure and dependencies. When unit testing controller logic, only the contents of a single action should be tested, not the behavior of its dependencies or of the framework itself. As you unit test your controller actions, make sure you focus only on its behavior. Don't unit test global or attribute-based :doc:`filters <filters>`, :doc:`routing </fundamentals/routing>`, or :doc:`model binding </mvc/models/model-binding>`, as part of your action's unit tests. These are all performed by the framework, and testing how they interact with your action is an integration test responsibility. 
```

Comment:
> When unit testing controller logic, only the contents of a single action should be tested

When unit testing controller logic, only the contents of a single action is tested